### ISSUE TITLE
User-created Mayastor Pool CRDs have missing "STATE" information

### ISSUE THREAD

**Issue by GlennBullingham** [openebs/Mayastor#419]
**Describe the bug**
When Mayastor Pool CRDs are created by the user, the corresponding pools are not actually created by Mayastor.

The output of ```kubectl -n mayastor get msp``` doesn't report a value for the STATE field of the affected pool(s)
e.g.
```
NAME         NODE                                STATE   AGE
diskpool-1   aks-agentpool-41030923-vmss000003           46m
diskpool-2   aks-agentpool-41030923-vmss000004           46m
diskpool-3   aks-agentpool-41030923-vmss000005           46m
```

The status section of any affected MSP CRD object is missing:

```
kubectl -n mayastor describe msp diskpool-1

Name:         diskpool-1
Namespace:    mayastor
Labels:       <none>
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Kind:         MayastorPool
Metadata:
  Creation Timestamp:  2020-09-16T15:35:55Z
  Generation:          1
  Resource Version:    13783
  Self Link:           /apis/openebs.io/v1alpha1/namespaces/mayastor/mayastorpools/diskpool-1
  UID:                 ba8adfd2-58a7-4fef-bd7f-0a24510157b1
Spec:
  Disks:
    /dev/sdc
  Node:  aks-agentpool-41030923-vmss000003
```

When this occurs, examination of the logs from the moac container and the associated mayastor container(s) will show that no messages regarding pool operations are logged - it appears as if MOAC has simply not seen, or ignored, the new pool CRD(s).

If the MOAC container is restarted (e.g. scale the moac deployment to 0 replicas and back to 1), the affected pool(s) are created and the corresponding MSP CRDs then present as expected (with pool State=online and their status section present).

**To Reproduce**

**This behaviour is intermittent** When it does manifest, it is necessary only to ```kubectl create``` a valid MSP CRD is order to observe the aberrant response (i.e. pool not created and MSP CRD lacking the Status section). 

**Expected behavior**

When a valid MSP CRD is created, the corresponding pool should be created by MOAC/Mayastor and report its state as online.

**OS info:**
 - Distro: Ubuntu 16.04.7 LTS (GNU/Linux 4.15.0-1092-azure x86_64) 
 - Kernel version: 4.15.0-1092-azure
 - MayaStor revision or container image: Mayastor v0.3.0, Mayastor RC v0.4.0

**Additional context**

**LOGS**
[moac_container_log.txt](https://github.com/openebs/Mayastor/files/5233917/moac_container_log.txt)
[mayastor_container_log.txt](https://github.com/openebs/Mayastor/files/5233918/mayastor_container_log.txt)

**Comment by paulyoong** [comment]
So far, I have been unable to reproduce this issue. Whenever I have created a pool, the status field of the MSP CRD has always been updated correctly (albeit this can take some time).

Having looked at _moac_container_log.txt_ there are a couple of lines that stands out:
`Sep 16 15:04:26.658 error [watcher]: Failed to get list of volume objects: HTTP 404`
`Sep 16 15:04:26.660 debug [watcher]: volume watch stream disconnected`

However, the volume watcher does subsequently connect:
`Sep 16 15:04:27.758 info [watcher]: volume watcher sync completed`

It is unclear if this is related to the issue. For one, the error is related to the volume watcher and not the pool/node watcher. Also, I have seen one case where a 404 error was returned and the pool status was still updated correctly.

At this stage more information is required to determine what is going on. Increasing the verbosity of the MOAC logging would be beneficial here as this will provide greater insight into the situation. This can be done by modifying _moac-deployment.yaml_ so that a "-vv" argument is supplied to the MOAC container, as shown below:
![image](https://user-images.githubusercontent.com/13060940/93903361-74ca7f80-fcf0-11ea-89d4-8d89fbc0c40f.png)

**Comment by tiagolobocastro** [comment]
@GlennBullingham looking at the CRD creation timestamp, looks like when the pool was created moac was already running. May I ask if this was indeed the case? If so then this issue is likely not related to the PR #392.

**Comment by GlennBullingham** [comment]
@tiagolobocastro Yes, that's very likely given the time of day concerned.

**Comment by jkryl** [comment]
I'm working on watcher rewrite just now and since it changes the whole watcher code there is a high probability that this could be fixed as part of it as well. Once it is pushed we will have to verify if the error is still reproducible.

**Comment by GlennBullingham** [comment]
I've just encountered this again, while testing RC 0.5.0 on Azure AKS.  I've attached the corresponding moac log, in which this time, there do appear indications that moac is hitting issues:

`Oct 12 20:02:27.515 error [node-operator]: Failed to update status of node resource "aks-agentpool-13453465-vmss000002": Error: Operation cannot be fulfilled on mayastornodes.openebs.io "aks-agentpool-13453465-vmss000002": the object has been modified; please apply your changes to the latest version and try again`

**Comment by jkryl** [comment]
Indeed, in this case, the root cause seems to be clear. We try to use an object from cache that hasn't been updated yet after previous create/modify operation. That is something that the new watcher implementation should prevent.

**Comment by jkryl** [comment]
The new watcher implementation that should prevent this from happening has been released in v0.6.
